### PR TITLE
[mlir] fix LLVM type converter for structs

### PR DESCRIPTION
**Pull request by ftynse**
Existing implementation of the LLVM type converter for LLVM structs containing incompatible types was attempting to change identifiers of the struct in case of name clash post-conversion (all identified structs have different names post-conversion since one cannot change the body of the struct once initialized). Beyond a trivial error of not updating the counter in renaming, this approach was broken for recursive structs that can't be made aware of the renaming and would use the pre-existing struct with clashing name instead.

For example, given

`!llvm.struct<"_Converted.foo", (struct<"_Converted.foo">, f32)>`

the following type

`!llvm.struct<"foo", (struct<"foo", index>)>`

would incorrectly convert to

```
!llvm.struct<"_Converted_1.foo",
             (struct<"_Converted.foo",
	             (struct<"_Converted.foo">, f32)>)>
```

Remove this incorrect renaming and simply refuse to convert types if it would lead to identifier clashes for structs with different bodies. Document the expectation that such generated names are reserved and must not be present in the input IR of the converter. If we ever actually need to use handle such cases, this can be achieved by temporarily renaming structs with reserved identifiers to an unreserved name and back in a pre/post-processing pass that does _not_ use the type conversion infra.

### CODE REVIEW

File: mlir/docs/TargetLLVMIR.md [hunk 1]
```diff
@@ -290,6 +290,21 @@ memref<2 x vector<4x8 x f32>
 Tensor types cannot be converted to the LLVM dialect. Operations on tensors must
 be [bufferized](Bufferization.md) before being converted.
 
+### Conversion of LLVM Container Types with Non-Compatible Element Types
+
+Progressive lowering may result in there existing LLVM container types, such
```
Comment (gysit) [inline mlir/docs/TargetLLVMIR.md:295]:
```suggestion
Progressive lowering may result in existing LLVM container types, such
```
nit: I believe the there should not be there?

File: mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp [hunk 2]
```diff
@@ -110,10 +102,26 @@ LLVMTypeConverter::LLVMTypeConverter(MLIRContext *ctx,
...
+          return failure();
+        }
+        results.push_back(convertedType);
+        return success();
+      }
+
+      // If it has been initialized and has the same body, just use it. This
+      // ensures that recursive structs keep being recursive rather than
+      // including a non-updated name.
+      if (TypeRange(convertedType.getBody()) == TypeRange(convertedElemTypes)) {
```
Comment (gysit) [inline mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp:119]: Question: Should the isPacked flag be checked here as well?

**Review by gysit — APPROVED** [review]
LGTM

It could be that the isPacked flag needs to be checked as well to avoid incompatibilities there?

File: mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp (hunk 2 above)
Comment (ftynse) [reply mlir/lib/Conversion/LLVMCommon/TypeConverter.cpp:119]: Good point, updating!

File: mlir/docs/TargetLLVMIR.md (hunk 1 above)
Comment (Dinistro) [reply mlir/docs/TargetLLVMIR.md]: You accidentally deleted the wrong word, I think.

**Pull request merged**